_kbarticlespages/single-page-application.md

```diff
@@ -0,0 +1,108 @@
+
+```
+const router = createBrowserRouter(
+  createRoutesFromElements(
+    // set route as the path plus layout
+    <Route path={path} element={<Layout />}>
+      <Route index element={<Home />}/>
+      <Route path={path + "Stuff"} element={<Stuff />}/>
+      <Route path={path + "contact"} element={<Contact />}/>
+      <Route path={path + "/*"} element={<Page404 />}/>
+    </Route>
+  )
+);
+
+createRoot(document.getElementById("root")).render(
+  <React.StrictMode>
+    <RouterProvider router={router} />
+  </React.StrictMode>
+);
+
+```
+
+
+
+Additionally within the `Layout` component is where we house all the `NavLinks` and `navbar` which are located at the top of the page. The `Outlet` component tells the `react router` where to output the child route page components within the layout. 
+
+```
+class Layout extends Component {
+
+    render() {
+        return (
+            <div>
+                <h1>Simple SPA</h1>
+                <ul className="header">
+                    {/* All nav links need to go to {path} */}
+                    <li><NavLink to={path}>Home</NavLink></li>
+                    <li><NavLink to={path + "stuff"}>Stuff</NavLink></li>
+                    <li><NavLink to={path + "contact"}>Contact</NavLink></li>
+                </ul>
+                <div className="content">
+                    <Outlet />
+                </div>
+            </div>
+        )
+    }
+}
+
+```
+
+It’s important to remember that none of these routes are creating new HTML files at a given path, they are only rendering in the browser (unless you prerender with [react-snap](https://github.com/stereobooster/react-snap)). We’ll see how to handle situations when the URL doesn’t match the HTML file in a later section.
+
+## Environment variables
+
+Pages provides [certain environment variables](https://cloud.gov/pages/documentation/env-vars-on-pages-builds/) in each build container and you can also add customer variables in your “Site Settings”. In the case of `create-react-app`, the library we’re using in this example, [we need two variables](https://create-react-app.dev/docs/advanced-configuration/):
+    - `BUILD_PATH`: this determines the output directory where all the built assets end up after running `react-scripts build`. In our case, Pages requires all assets to be in the `_site` folder. We can set that in the UI like so:
+<img src="{{ site.baseurl }}/assets/images/pages/env_var.png"/>
+
+`PUBLIC_URL`: this variable determines how assets are referenced in the final build. We can’t set this one statically in the UI because our root URL will change for preview builds. Instead, we’ll add a small bash script to our custom build command to set the default environment variable `BASEURL` as the value for `PUBLIC_URL` like this:
+
+```
+echo "PUBLIC_URL=$BASEURL" >> .env
+```
+	
+  You can see how this is included in the build command in the [repository](https://github.com/Ephraim-G/react_spa4/blob/main/build.sh#L3C32-L3C32). 
```

Comment:
You've got a `postbuild.sh` in the pages build script in package.json that does not exist and fails the build. https://github.com/Ephraim-G/react_spa4/blob/main/package.json#L21